src/sstruct_mv/HYPRE_sstruct_matrix.c

```diff
@@ -617,6 +618,76 @@ HYPRE_SStructMatrixGetBoxValues2( HYPRE_SStructMatrix  matrix,
    return hypre_error_flag;
 }
 
+/*--------------------------------------------------------------------------
+ *--------------------------------------------------------------------------*/
+
+HYPRE_Int
+HYPRE_SStructMatrixAddFEMBoxValues(HYPRE_SStructMatrix  matrix,
+                                   HYPRE_Int            part,
+                                   HYPRE_Int           *ilower,
+                                   HYPRE_Int           *iupper,
+                                   HYPRE_Complex       *values)
+{
+   HYPRE_Int             ndim            = hypre_SStructMatrixNDim(matrix);
+   hypre_SStructGraph   *graph           = hypre_SStructMatrixGraph(matrix);
+   hypre_SStructGrid    *grid            = hypre_SStructGraphGrid(graph);
+   HYPRE_MemoryLocation  memory_location = hypre_IJMatrixMemoryLocation(
```

Comment:
Can we use `hypre_SStructMatrixMemoryLocation`? Is it better to use?

Reply:
We can't because it returns `HYPRE_MEMORY_UNDEFINED` if called prior to the matrix assembly. However, I think it's more appropriate to do as you said.

To make that work, we would need to move
https://github.com/hypre-space/hypre/blob/467e3f3d9002fae4b71f46b07876884a23a8e7e0/src/sstruct_mv/sstruct_matrix.c#L1301-L1302 
from `hypre_SStructUMatrixAssemble` to after the line

https://github.com/hypre-space/hypre/blob/467e3f3d9002fae4b71f46b07876884a23a8e7e0/src/sstruct_mv/sstruct_matrix.c#L778
which is at `hypre_SStructUMatrixInitialize`. Then, `hypre_SStructMatrixMemoryLocation` would return the correct memory location when called prior to assembling the matrix. Let's do this change?

Reply:
Got it. Good point. Not sure about the changes you suggested, which seem to be big. I am not as familiar with SStruct, so I will rely on the expert decisions from others. The current code looks fine to me though. Thank you!

Reply:
@rfalgout can you comment on my suggestion above? It seems safe to move the GetObject call to right after initialize. This change will allow using `HYPRE_SStructMatrixMemoryLocation` in the new BoxValues routines.

Reply:
Yes, this is the right place to put it anyway I think.  Thanks!

Reply:
>It seems safe to move the GetObject call to right after initialize

Done in [5138246](https://github.com/hypre-space/hypre/pull/779/commits/5138246d287c1513edc75759d7c2cf111daa6d2e)